File: server/paths.js
'use strict';

const path = require('path');

const ROOT_DIR = path.resolve(__dirname, '..');

const PATHS = Object.freeze({
  root: ROOT_DIR,
  settings: path.join(ROOT_DIR, 'appsettings.json'),
  settingsSample: path.join(ROOT_DIR, 'appsettings.sample.json'),
  tickets: path.join(ROOT_DIR, 'tickets.json'),
  drawState: path.join(ROOT_DIR, 'winners.json'),
});

// Files that must never be reachable through the static file middleware.
// appsettings.json carries the hashed admin credentials; winners.json carries
// participant contact details of everyone already drawn, and tickets.json those
// of everyone entered.
const PROTECTED_FILES = Object.freeze([
  'appsettings.json',
  'winners.json',
  // FIX: tickets.json was served as a static file, so the whole entry list —
  // names, companies and the columns marked sensitive, such as phone numbers —
  // could be downloaded by anyone who guessed the URL. The board never needs
  // it: it reads entries through /api/pool, which strips non-display fields.
  'tickets.json',
  'package.json',
  'package-lock.json',
  'requirements.txt',
]);

const PROTECTED_DIRS = Object.freeze(['server', 'pyserver', 'windows', '.git', 'node_modules', '.playwright-mcp']);

function isProtectedPath(requestPath) {
  const normalized = String(requestPath || '')
    .replace(/^\/+/, '')
    .split('?')[0]
    .toLowerCase();

  if (!normalized) return false;
  if (PROTECTED_FILES.some((file) => normalized === file.toLowerCase())) return true;

  const firstSegment = normalized.split('/')[0];
  return PROTECTED_DIRS.includes(firstSegment);
}

module.exports = { PATHS, PROTECTED_FILES, PROTECTED_DIRS, isProtectedPath };

File: pyserver/paths.py
"""Filesystem locations and the deny-list used by the static file route."""

from pathlib import Path

ROOT_DIR = Path(__file__).resolve().parent.parent

SETTINGS_PATH = ROOT_DIR / "appsettings.json"
SETTINGS_SAMPLE_PATH = ROOT_DIR / "appsettings.sample.json"
TICKETS_PATH = ROOT_DIR / "tickets.json"
DRAW_STATE_PATH = ROOT_DIR / "winners.json"

# appsettings.json holds the hashed admin credentials, winners.json the contact
# details of everyone drawn so far and tickets.json those of everyone entered;
# none of them may be served as a static asset.
PROTECTED_FILES = (
    "appsettings.json",
    "winners.json",
    # FIX: tickets.json was served as a static file, so the whole entry list —
    # names, companies and the columns marked sensitive, such as phone numbers —
    # could be downloaded by anyone who guessed the URL. The board never needs
    # it: it reads entries through /api/pool, which strips non-display fields.
    "tickets.json",
    "package.json",
    "package-lock.json",
    "requirements.txt",
)

PROTECTED_DIRS = ("server", "pyserver", "windows", ".git", "node_modules", ".playwright-mcp")


def is_protected_path(request_path: str) -> bool:
    normalized = (request_path or "").lstrip("/").split("?")[0].lower()
    if not normalized:
        return False
    if normalized in {name.lower() for name in PROTECTED_FILES}:
        return True
    return normalized.split("/")[0] in PROTECTED_DIRS
